pkg/cmd/render/render.go

```diff
@@ -181,142 +177,110 @@ func (r *renderOpts) Run() error {
-		if err != nil {
-			return nil
-		}
-		// TODO make me pretty and a function for ez testing
-		if singleStack {
-			// IPv6
-			etcdAddress := options.EtcdAddress{
-				ListenClient:       net.JoinHostPort(net.IPv6zero.String(), "2379"),
-				ListenPeer:         net.JoinHostPort(net.IPv6zero.String(), "2380"),
-				LocalHost:          "[::]",
-				ListenMetricServer: net.JoinHostPort(net.IPv6zero.String(), "9978"),
-				ListenMetricProxy:  net.JoinHostPort(net.IPv6zero.String(), "9979"),
-			}
-			renderConfig.ManifestConfig.EtcdAddress = etcdAddress
-		} else {
-			// IPv4
-			etcdAddress := options.EtcdAddress{
-				ListenClient:       net.JoinHostPort(net.IPv4zero.String(), "2379"),
-				ListenPeer:         net.JoinHostPort(net.IPv4zero.String(), "2380"),
-				LocalHost:          "127.0.0.1",
-				ListenMetricServer: net.JoinHostPort(net.IPv4zero.String(), "9978"),
-				ListenMetricProxy:  net.JoinHostPort(net.IPv4zero.String(), "9979"),
-			}
-			renderConfig.ManifestConfig.EtcdAddress = etcdAddress
-		}
+
+	if err := templateData.setClusterCIDR(opts.clusterConfigFile); err != nil {
+		return nil, err
+	}
+	if err := templateData.setServiceCIDR(opts.clusterConfigFile); err != nil {
+		return nil, err
+	}
+	if err := templateData.setSingleStackIPv6(); err != nil {
+		return nil, err
 	}
+	templateData.setEtcdAddress()
+
+	return &templateData, nil
+}
 
-	if err := r.manifest.ApplyTo(&renderConfig.ManifestConfig); err != nil {
+// Run contains the logic of the render command.
+func (r *renderOpts) Run() error {
+	templateData, err := newTemplateData(r)
+	if err != nil {
+		return err
+	}
+
+	if err := r.manifest.ApplyTo(&templateData.ManifestConfig); err != nil {
 		return err
 	}
 	if err := r.generic.ApplyTo(
-		&renderConfig.FileConfig,
+		&templateData.FileConfig,
 		options.Template{FileName: "defaultconfig.yaml", Content: v430_00_assets.MustAsset(filepath.Join(bootstrapVersion, "etcd", "defaultconfig.yaml"))},
 		mustReadTemplateFile(filepath.Join(r.generic.TemplatesDir, "config", "bootstrap-config-overrides.yaml")),
 		mustReadTemplateFile(filepath.Join(r.generic.TemplatesDir, "config", "config-overrides.yaml")),
-		&renderConfig,
+		&templateData,
 		nil,
 	); err != nil {
 		return err
 	}
 
-	return WriteFiles(&r.generic, &renderConfig.FileConfig, renderConfig)
+	return WriteFiles(&r.generic, &templateData.FileConfig, templateData)
 }
 
-// parseStaticTemplateFile takes a path to a yaml template file returns a populated File struct.
-func parseStaticTemplateFile(path string) (*File, error) {
-	data, err := ioutil.ReadFile(path)
-	if err != nil {
-		return nil, err
-	}
+func (t *TemplateData) setEtcdAddress() {
+	// IPv4
+	allAddresses := "0.0.0.0"
+	localhost := "127.0.0.1"
 
-	file := new(File)
-	if err := yaml.Unmarshal(data, file); err != nil {
-		return nil, fmt.Errorf("failed to unmarshal file into struct: %v", err)
+	// IPv6
+	if t.SingleStackIPv6 {
+		allAddresses = "[::]"
+		localhost = "[::1]"
```

Comment:
You're using `net.JoinHostPort` below, so the brackets here are redundant (and result in eventually generating `--listen-client-urls https://[[::]]:2379`)

Reply:
nice, thanks